ckanext/dcat/utils.py

```diff
@@ -0,0 +1,146 @@
+import logging
+
+from pylons import config
+
+from ckan import model
+
+log = logging.getLogger(__name__)
+
+
+def catalog_uri():
+    '''
+    Returns an URI for the whole catalog
+
+    This will be used to uniquely reference the CKAN instance on the RDF
+    serializations and as a basis for eg datasets URIs (if not present on
+    the metadata).
+
+    The value will be the first found of:
+
+        1. The `ckanext.dcat.base_uri` config option (recommended)
+        2. The `ckan.site_url` config option
+        3. `http://` + the `app_instance_uuid` config option (minus brackets)
+
+    A warning is emited if the third option is used.
+
+    Returns a string with the catalog URI.
+    '''
+
+    uri = config.get('ckanext.dcat.base_uri')
+    if not uri:
+        uri = config.get('ckan.site_url')
+    if not uri:
+        uri = 'http://' + config.get('app_instance_uuid').replace('{', '') \
+                                                         .replace('}', '')
+        log.critical('Using app id as catalog URI, you should set the ' +
+                     '`ckanext.dcat.base_uri` or `ckan.site_url` option')
+    return uri
+
+
+def dataset_uri(dataset_dict):
+    '''
+    Returns an URI for the dataset
+
+    This will be used to uniquely reference the dataset on the RDF
+    serializations.
+
+    The value will be the first found of:
+
+        1. The value of the `uri` field
+        2. The value of an extra with key `uri`
+        3. `catalog_uri()` + '/dataset/' + `id` field
```

Comment:
@davidread I can see where you are coming from, I just feel that specially when dealing when harvested datasets coming from different sources you really want that to be immutable. +1 on getting the opinion of an expert if we can find one :)